core/src/bot/dto/play.dto.ts

```diff
@@ -9,5 +9,5 @@ export class PlayDto {
       'Name or URL of song/playlist. Could be from (Youtube, Spotify, SoundCloud)',
     required: true,
   })
-  song: string;
+  song!: string;
```

Comment:
What does the bang do here? First time seeing this.

Reply:
[Definite assignment](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#definite-assignment-assertions).

basically, we're defining these properties, but there's no code that gives them a value explicitly. if they were  `| undefined`, it wouldn't have an issue with this, as it would assume they're defaulting to `undefined`. but because they're not, it's looking for an assignment and erroring out because they're never set (this is a strict null checks thing).

By using `!`, we're telling the compiler that we know more than it does, and that despite there being no assignment, it will always be present anyway. as devs, we know this already because of the semantics of the decorators